File: 300-longest-increasing-subsequence/longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> tails; // tails[i] = smallest possible tail value of an increasing subsequence of length i+1
        
        for (int num : nums) {
            // Binary search for the leftmost position where num can replace/extend
            auto it = lower_bound(tails.begin(), tails.end(), num);
            
            if (it == tails.end()) {
                tails.push_back(num); // num extends the longest subsequence found so far
            } else {
                *it = num; // num replaces an element to keep tails as small as possible
            }
        }
        
        return tails.size();
    }
};

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        tails = []  # tails[i] = smallest possible tail value of an increasing subsequence of length i+1
        
        for num in nums:
            # Binary search for the leftmost position where num can replace/extend
            lo, hi = 0, len(tails)
            while lo < hi:
                mid = (lo + hi) // 2
                if tails[mid] < num:
                    lo = mid + 1
                else:
                    hi = mid
            
            if lo == len(tails):
                tails.append(num)  # num extends the longest subsequence found so far
            else:
                tails[lo] = num    # num replaces an element to keep tails as small as possible
        
        return len(tails)